### PR TITLE
[fix](fold) fixed an issue with be computing constants

### LINKED ISSUE #43061: [Bug] enable_fold_constant_by_be设置为true和false的时候sql结果不一致
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Version

2.1.7

### What's Wrong?

mysql> set enable_fold_constant_by_be = true;
Query OK, 0 rows affected (0.00 sec)

mysql> select hex(s) from ( select from_base64('wr2JEDVXzL9+2XtRhgIloA==') as s )t;
+--------------------------------------------------+
| hex(s)                                                                   |
+--------------------------------------------------+
| C2BDEFBFBD103557CCBF7EEFBFBD7B51EFBFBD0225EFBFBD |
+--------------------------------------------------+
1 row in set (0.02 sec)

mysql> set enable_fold_constant_by_be = false;
Query OK, 0 rows affected (0.00 sec)

mysql> select hex(s) from ( select from_base64('wr2JEDVXzL9+2XtRhgIloA==') as s )t;
+----------------------------------+
| hex(s)                           |
+----------------------------------+
| C2BD89103557CCBF7ED97B51860225A0 |
+----------------------------------+
1 row in set (0.01 sec)

### What You Expected?

设置set enable_fold_constant_by_be前后结果查询应该一致

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by felixwluo**
### What problem does this PR solve?
issue close: #43061

1、Problem
`When enable_fold_constant_by_be=true is set, the results of select hex(from_base64('wr2JEDVXzL9+2XtRhgIloA==')) and select hex(s) from (select from_base64('wr2JEDVXzL9+2XtRhgIloA==') as s) t are inconsistent`

2、mysql results
```
select hex(s) from (select from_base64('wr2JEDVXzL9+2XtRhgIloA==') as s) t;
C2BD89103557CCBF7ED97B51860225A0
```

3、cause
`When processing binary data such as FromBase64, BE will return the original binary data through the bytesValue field, and the previous code only uses the stringValue field, resulting in the binary data being corrupted during the string encoding conversion process`

### CODE REVIEW

**Comment by felixwluo** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 1]
```diff
@@ -263,6 +264,16 @@ private static boolean shouldSkipFold(Expression expr) {
...
 
+        if (expr instanceof FromBase64) {
+            return true;
+        }
+
+        for (Expression child : expr.children()) {
+            if (shouldSkipFold(child)) {
+                return true;
+            }
+        }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:275]: should not add this, check children will be done out side

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 2]
```diff
@@ -263,6 +264,16 @@ private static boolean shouldSkipFold(Expression expr) {
             return true;
         }
 
+        if (expr instanceof FromBase64) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:267]: why fromBase64 should not folding?

Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:267]: u only explain the result is not right, but not explain why

**Comment by felixwluo** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java [hunk 3]
```diff
@@ -486,9 +486,27 @@ public static List<Literal> getResultExpression(DataType type, PValues resultCon
...
+                        res.add(new NullLiteral(type));
+                    } else {
+                        res.add(new StringLiteral(stringValue));
+                    }
+                }
+            } else {
+                num = resultContent.getStringValueCount();
+                for (int i = 0; i < num; ++i) {
+                    String stringValue = resultContent.getStringValue(i);
+                    if ("\\N".equalsIgnoreCase(stringValue) && resultContent.hasHasNull()) {
```
Comment (LiBinfeng-01) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/FoldConstantRuleOnBE.java:504]: we could extract this replicate dealing of string to a seperate function to reduce codes

**Comment by LiBinfeng-01** [comment]
pls explain how this problem was fixed, and leave some comments

**Comment by coder-coding-the-code** [comment]
I think you should use explain to see if constant folding still exists

**Comment by felixwluo** [comment]
run buildall

**Comment by felixwluo** [comment]
run feut

**Comment by felixwluo** [comment]
run buildall

**Review by morrySnow — DISMISSED** [review]

**Review by LiBinfeng-01 — APPROVED** [review]

**Comment by felixwluo** [comment]
run p0

**Comment by felixwluo** [comment]
run feut

**Comment by felixwluo** [comment]
run buildall

**Comment by felixwluo** [comment]
run p0

**Comment by felixwluo** [comment]
run feut

**Review by morrySnow — APPROVED** [review]

**Pull request merged**